### ISSUE TITLE
When enabledProtocols are empty, use the proper default client protocols

### ISSUE THREAD

**Issue by yiyangc91** [AsyncHttpClient/async-http-client#991]
In both 1.9.x and master (haven't checked 1.8.x) the SSLEngineFactory has this bit of code:

```java
             SSLEngine sslEngine = sslContext.createSSLEngine(peerHost, peerPort);
             if (!config.isAcceptAnyCertificate()) {
                 SSLParameters params = sslEngine.getSSLParameters();
                 params.setEndpointIdentificationAlgorithm("HTTPS");
                 sslEngine.setSSLParameters(params);
             }
            sslEngine.setUseClientMode(true);
```

Unfortunately running setSSLParameters before setUseClientMode prevents SSLEngine from replacing the default server-mode protocols with the default client-mode protocols.

This means that the default SSL Client Hello used is SSLv2Hello, as opposed to SSLv3, which does not play nicely with SNI.

The fix is to merely move the "setUseClientMode" line above the "setSSLParameters".

**Comment by yiyangc91** [comment]
Raised PRs #992 and #993

**Comment by slandelle** [comment]
> Unfortunately running setSSLParameters before setUseClientMode prevents SSLEngine from replacing the default server-mode protocols with the default client-mode protocols.

Could you elaborate, please? I checked [sun.security.ssl.SSLEngineImpl source code](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/sun/security/ssl/SSLEngineImpl.java), and it looks to me that, as at this point `connectionState` is set to `cs_START` and `handshaker` is `null`, it doesn't change anything.

**Comment by slandelle** [comment]
Which JDK do you use?

**Comment by slandelle** [comment]
What you describe would happen if `connectionState` is `cs_HANDSHAKE` and `handshaker` is not `null`. I don't get how this could happen in our usage. Could you please explain how you got there?

**Comment by yiyangc91** [comment]
Clarification:

1. ```createSSLEngine``` creates an ```SSLEngine```, which calls ```private void init(SSLContextImpl ctx)```. This creates an SSLEngine with ```enabledProtocols = sslContext.getDefaultProtocolList(roleIsServer);```, where roleIsServer is always true. The default server protocols for the DefaultSSLContext is SSLv2Hello, SSLv3, TLSv10, TLSv11 and TLSv12
2. ```sslEngine.getSSLParameters()``` pulls the list of enabled protocols out as a list of strings.
3. We then call ```sslEngine.setSSLParameters(params);```, we've eventually call ```setSSLParameters(SSLParameters params)``` in [[http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/javax/net/ssl/SSLEngine.java]], which creates a *new* ```ProtocolList```, containing the list of strings.
4. We then call ```sslEngine.setUseClientMode(true);```, which checks ```this.sslContext.isDefaultProtocolList(this.enabledProtocols))```, which does a compare-by-reference of the enabledProtocols ProtocolList. This is the issue! Because we've replaced the default ```enabledProtocols``` with a *new ProtocolList*, the reference comparison fails, and we fail to replace the *default server-mode enabledProtocols* with the *default client-mode enabledProtocols*.

I am using Java 1.8.0_60

**Comment by slandelle** [comment]
We [explicitly set the enabled protocols](https://github.com/AsyncHttpClient/async-http-client/blob/master/client/src/main/java/org/asynchttpclient/channel/SSLEngineFactory.java#L63).

> The default client mode protocols are SSLv3, TLSv10, TLSv11 and TLSv12.

No, default enabled protocols in AHC are [TLSv12, TLSv11 and TLSv10](https://github.com/AsyncHttpClient/async-http-client/blob/master/client/src/main/resources/ahc-default.properties#L14). SSLv3 is disabled by default.

> Set up an Apache Hostname based VirtualHost that has a different SSL certificate to the one in the default domain

I reaaallly don't have time to set this up, sorry. I'd gladly test if you could provide a running server.

**Comment by yiyangc91** [comment]
I didn't realize 2.x had a default properties that was used for configuration. This bug doesn't affect master in that case.

The TL;DR is that you've constructed an SSLEngine with enabledProtocols as ```this.sslContext.getDefaultProtocolList(true);```. Then you **new**'d all over the enabledProtocols in the SSLEngineImpl, so when it comes time to setUseClientMode(true), the reference has changed, and thus it doesn't bother calling ``` this.sslContext.getDefaultProtocolList(false);``` to replace the existing defaults.

You really want to allow the default client-mode protocols to be used instead, and then you can trample over the reference.

This only happens when Ning has not set any enabledProtocols in the config (which happens in 1.9.x). Maybe you can decide that 1.9.x isn't supported anymore (shrug).

>No, default enabled protocols in AHC are TLSv12, TLSv11 and TLSv10. SSLv3 is disabled by default.

But the defaults in DefaultSSLContext *is* what I mentioned, which is being used in 1.9.x.

**Comment by yiyangc91** [comment]
I don't have an Apache set up either ATM. You could probably use something like sni.velox.ch, which will balk if you don't send servername.

**Comment by slandelle** [comment]
> You really want to allow the default client-mode protocols to be used instead, and then you can trample over the reference.

We can change, it would be cleaner, but I really don't think there's an issue on AHC side. SSLv3 is disabled by default, even in 1.9.

> This only happens when Ning has not set any enabledProtocols in the config (which happens in 1.9.x). 

Nope.
Since 1.9.30, SSLv3 is disabled by default: https://github.com/AsyncHttpClient/async-http-client/issues/934
I suspect you AHC version is simply not up-to-date, latest is 1.9.31.

**Comment by yiyangc91** [comment]
Oh okay, yeah actually you're correct. I'm using scala-dispatch and that's pulled in 1.9.11. I suppose I should manually bump the version and raise an issue on scala-dispatch.

I think this is more of a cleanliness issue then rather than a bug then. If the user goes and nulls the enabledProtocols, it could probably happen again.

Thanks for your help in resolving this.

**Comment by slandelle** [comment]
> I think this is more of a cleanliness issue

Agree. I'll rename the issue.

> Thanks for your help in resolving this.

You're welcome
